geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java

```diff
@@ -422,16 +424,21 @@ synchronized boolean membersRemove(byte[] memberToRemove) {
    * @return the number of members actually added
    */
   public long sadd(List<byte[]> membersToAdd, Region<RedisKey, RedisData> region, RedisKey key) {
-    AddByteArrays delta = new AddByteArrays();
+    AddByteArrays delta;
+    byte newVersion;
     int membersAdded = 0;
-    for (byte[] member : membersToAdd) {
-      if (membersAdd(member)) {
-        delta.add(member);
-        membersAdded++;
+    synchronized (this) {
+      newVersion = incrementAndGetVersion();
+      delta = new AddByteArrays(newVersion);
+      for (byte[] member : membersToAdd) {
+        if (membersAdd(member)) {
+          delta.add(member);
+          membersAdded++;
+        }
+      }
+      if (membersAdded == 0) {
```

Comment:
The sync can end before you check if membersAdded is 0

Reply:
No longer touching RedisSet after rebase